File: modules/core/src/main/java/turbolift/internals/engine/concurrent/atomic/AtomicLongVH.java
package turbolift.internals.engine.concurrent.atomic;
import java.lang.invoke.VarHandle;
import java.lang.invoke.MethodHandles;


public class AtomicLongVH {
  private long valueVH;
  private static final VarHandle nameVH;

  static {
    try {
      nameVH = MethodHandles.lookup().findVarHandle(AtomicLongVH.class, "valueVH", long.class);
    } catch (ReflectiveOperationException e) {
      throw new ExceptionInInitializerError(e);
    }
  }

  public AtomicLongVH(long a) {
    valueVH = a;
  }

  protected long getVH() {
    return (long) nameVH.get(this);
  }

  protected void setVH(long a) {
    nameVH.set(this, a);
  }

  protected long gasVH(long a) {
    return (long) nameVH.getAndSet(this, a);
  }

  protected boolean casVH(long a, long b) {
    return nameVH.compareAndSet(this, a, b);
  }

  protected long caxVH(long a, long b) {
    return (long) nameVH.compareAndExchange(this, a, b);
  }

  protected long gaaVH(long d) {
    return (long) nameVH.getAndAdd(this, d);
  }
}

File: modules/core/src/main/scala/turbolift/internals/engine/concurrent/atomic/AtomicLongVH.scala
package turbolift.internals.engine.concurrent.atomic
import java.lang.invoke.{VarHandle, MethodHandles}


private[turbolift] class AtomicLongVH(protected var unsafeValue: Long):
  protected def getVH: Long = AtomicLongVH.theVH.get(this)
  
  protected def setVH(a: Long): Unit = AtomicLongVH.theVH.set(this, a)
  
  protected def gasVH(a: Long): Long = AtomicLongVH.theVH.getAndSet(this, a)

  protected def casVH(a: Long, b: Long): Boolean = AtomicLongVH.theVH.compareAndSet(this, a, b)

  protected def caxVH(a: Long, b: Long): Long = AtomicLongVH.theVH.compareAndExchange(this, a, b)


private[turbolift] object AtomicLongVH:
  private val theVH: VarHandle = MethodHandles
    .privateLookupIn(classOf[AtomicLongVH], MethodHandles.lookup()).nn
    .findVarHandle(classOf[AtomicLongVH], "unsafeValue", java.lang.Long.TYPE).nn
